x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/WordPieceTokenFilter.java

```diff
@@ -0,0 +1,208 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+
+public final class WordPieceTokenFilter extends TokenFilter {
+    protected final LinkedList<WordPieceToken> tokens;
+    private final CharTermAttribute termAtt = addAttribute(CharTermAttribute.class);
+    protected final OffsetAttribute offsetAtt = addAttribute(OffsetAttribute.class);
+    private final PositionIncrementAttribute posIncAtt = addAttribute(PositionIncrementAttribute.class);
+    private static final CharSequence CONTINUATION = "##";
+
+    private State current;
+    private final CharArraySet neverSplit;
+    private final CharArrayMap<Integer> vocabulary;
+    private final List<WordPieceToken> tokenizedValues;
+    private final int maxInputCharsPerWord;
+    private final int tokenizedUnknown;
+    private final CharSequence unknownToken;
+
+    public static WordPieceTokenFilter buildFromSettings(
+        boolean isLowerCase,
+        boolean isTokenizeCjkChars,
+        boolean isStripAccents,
+        List<String> neverSplit,
+        List<String> dictionary,
+        String unknownToken,
+        int maxInputCharsPerWord,
+        TokenStream input
+    ) throws IOException {
+        CharArrayMap<Integer> vocabMap = new CharArrayMap<>(dictionary.size(), isLowerCase);
+        int i = 0;
+        for (var word : dictionary) {
+            vocabMap.put(word, i++);
+        }
+        input = BasicTokenFilter.buildFromSettings(isLowerCase, isTokenizeCjkChars, isStripAccents, neverSplit, input);
+        return new WordPieceTokenFilter(input, new CharArraySet(neverSplit, isLowerCase), vocabMap, unknownToken, maxInputCharsPerWord);
+    }
+
+    public WordPieceTokenFilter(
+        TokenStream input,
+        CharArraySet neverSplit,
+        CharArrayMap<Integer> vocabulary,
+        CharSequence unknownToken,
+        int maxInputCharsPerWord
+    ) {
+        super(input);
+        this.tokens = new LinkedList<>();
+        this.neverSplit = neverSplit;
+        this.vocabulary = vocabulary;
+        this.tokenizedValues = new ArrayList<>();
+        if (vocabulary.containsKey(unknownToken) == false) {
+            throw new IllegalArgumentException(
+                "provided vocabulary does not contain the unknown token of [" + unknownToken.toString() + "]"
+            );
+        }
+        this.unknownToken = unknownToken;
+        this.tokenizedUnknown = vocabulary.get(unknownToken);
+        this.maxInputCharsPerWord = maxInputCharsPerWord;
+    }
+
+    public List<WordPieceToken> getTokenizedValues() {
+        return tokenizedValues;
+    }
+
+    @Override
+    public void reset() throws IOException {
+        super.reset();
+        tokens.clear();
+        tokenizedValues.clear();
+        current = null;
+    }
+
+    @Override
+    public boolean incrementToken() throws IOException {
+        // TODO seems like this + lowercase + tokenize cjk + wordpiece could all be the same thing....
```

Comment:
Yes, I agree.  The different filters are not being composed in the analyzer config, they're being composed within this token filter.  So they might as well all be part of this same token filter, which will mean fewer intermediate data structures.  The only reason for keeping them completely separate is if we wanted the building blocks to be reusable generally within Elasticsearch.  But Elasticsearch must already have similar functionality available.  Therefore, it's better that eventually we publicly expose "word_piece" and it has all these options built in.  Then there's no need to have several classes extending `TokenFilter` that are not registered as generally available token filters.

Reply:
@droberts195 I do think have the "basic" tokenizer separate from wordpiece makes sense as I think it is used by other tokenization schemes. 

But I can totally collapse the other things into `basic` to make it just two filters